### ISSUE TITLE
Preventing server-forced updates

### ISSUE THREAD

**Issue by mappum** [slightlyoff/ServiceWorker#822]
I'd like to propose a change to the ServiceWorker spec that allows for applications which cannot be forced by the server to install updates.

### Background

Native applications have a significant security advantage over web applications: they control their own update lifecycles. New code is not loaded by default, so if there is no auto-update functionality built in, new versions are only used at the user's will. By contrast, web applications load new, untrusted code from a server each time the user navigates to them. This increases the attack surface for malicious code to be deployed by whoever has control of the server (hackers, political actors, rogue service administrators, etc.).

The Application Cache API allowed this feature; servers could set the AppCache manifest `max-age` very far in the future which would prevent the browser from contacting the server. However, servers configured to cache aggressively would cache these manifests and brick apps accidentally, leaving developers unable to deploy fixes to their apps. Since this problem was common, the ServiceWorker spec is defined to limit `max-age` to 24 hours, after which an update is ensured to happen.

### Use Cases

Server-forced updates need to be prevented in the following cases:
- Handling private keys or other sensitive data
- Apps that rely on anonymity or encryption, such as TOR or secure messaging
- ServiceWorkers which verify integrity of updates loaded from an untrusted CDN

For an example of code that uses this ability of the Application Cache API, see [`substack/hyperboot`](https://github.com/substack/hyperboot) (written by @substack, the author of [`browserify`](https://github.com/substack/node-browserify)).

### Solutions

In the ServiceWorker spec, preventing forced updates must only happen when explicitly requested, so as not to cause the accidental bricking tragedies seen with AppCache.

Possible methods for adopting this feature:
- Removing the 24-hour cap for the proposed `Service-Worker-Max-Age` header (#721). Servers likely won't set this header by default.
- An additional `Service-Worker-No-Forced-Updates` header, that should only be used to specifically opt in to this feature, which will remove the 24-hour cap.

These changes still allow for applications to trigger their own updates by unregistering their ServiceWorker and registering the new version. It may also be beneficial to add a method which triggers a reload of the registered ServiceWorker and does the standard byte-for-byte comparison to see if there is an update.

### Thank You

Thanks for considering this proposal. I believe this small change would make ServiceWorkers much more powerful, and bring the web a huge step closer to parity with native applications.

**Comment by mappum** [comment]
BTW, it was a great idea by you guys to develop this spec on Github. :+1: This allows for a very open discussion. (I have never made any proposals like this to any sort of standards bodies before this).

**Comment by jakearchibald** [comment]
I can't quite get my dead around the use cases.

> Handling private keys or other sensitive data
> Apps that rely on anonymity or encryption, such as TOR or secure messaging

How does the current update model prevent this?

> ServiceWorkers which verify integrity and authenticity of updates loaded from an untrusted CDN

SW scripts are same-origin for security reasons. When you're adding thigns to the cache you can already verify integrity, although CSP is a better mechanism for this.

**Comment by mappum** [comment]
>I can't quite get my dead around the use cases.

Imagine we have created a PGP app to encrypt messages, and we serve it at https://mypgpapp.com. It generates private keys and stores them in IndexedDB, then lets the user encrypt messages with those keys so they can copy their encrypted messages to send via email.

So far, our security is pretty good. Since we used HTTPS, we can be reasonably sure the user won't load a backdoored script from some attacker. However, what happens if *our server* gets compromised? If an attacker had access to it, they could deploy an update to the registered SW script which fetches code that will upload the private key somewhere. Now it's game over for any user who visits the app and has their SW script caching expire.

>How does the current update model prevent this?

By this example I meant that when the app is being hosted on servers owned by some other party, the host who actually runs the machines could deploy backdoored code. But if a SW prevented forced updating, it could be built to check that the updated version is cryptographically signed by the author before accepting updates. Then, the author can sign the resources offline, and never has to give the private key to the host to have the users verify authenticity.

CSP doesn't solve this since the malicious scripts are coming from a whitelisted domain.

(I suppose I didn't need to mention integrity, authenticity was the important part).

**Comment by delapuente** [comment]
Related if not dup of #761

**Comment by asutherland** [comment]
I resonate with this use-case (I worked on the Firefox OS mail app which similarly wanted what amounted to signed-offline app packages), but I think giving the service-worker the ability to defeat upgrades of itself is the wrong way to handle it.  (And I think this issue is roughly equivalent to 761, although this issue more concretely describes motivating use-cases.)

The security model of https is trust in an origin, authenticated by the certificate authority (CA) infrastructure, with compromises being handled via OCSP (stapling) and/or short-lived certificates.  Using a trust-on-first-use service worker that can deny upgrades is a clever way to attempt to embed an offline signature based security/trust model inside the existing online signature trust model.  Some people have even come up with an exceedingly clever approach also combining HPKP and AppCache, see https://www.reddit.com/r/encryption/comments/4027ci/how_2_spacex_alums_are_using_encryption_for_good/cyywmc8

The major risk of a prevent-update feature, as covered by Jake, in #761 is allowing an evil/attacker-controlled service worker to extend the duration of its nefarious residency, potentially permanently.  And also as Jake suggests, it seems better to make the validation part of a standard already concerned with validating contents, namely CSP which already has precedent with its SRI-like "valid hash" mechanism https://w3c.github.io/webappsec-csp/2/#source-list-valid-hashes.  This also allows the browser to alert the user, helping avoid spoofable UI and related confusion/fatigue.

Additionally:
* There are potential interactions such as shift+reload bypassing service-workers, among others.  It's a lot to ask a complex spec like service-workers to adopt a fundamentally divergent use-case that in many cases is indistinguishable from an attacker's dream and a developer-who-makes-typos' nightmare.
* Preventing updates is a shallow security model on its own.  The origin is being defended exclusively by the service worker intercepting all requests and maintaining itself intact.  If an attacker does gain control of the origin and can cause the service worker to fail to respondWith or is bypassed via some other bug or new feature (like a bypass header), they have access to all of your app's local storage at the origin.  What you really want is your own distinct origin with its own storage bucket that is affirmatively guarded by verifying a signature before letting any code/document run in the origin.  Gecko has internal support for such things hung off an existing origin using OriginAttributes, or just if a separate scheme and thereby origin is created like "signed-package://KEY", etc.  This would have been surfaced by the now-defunct Firefox OS "new security model" at https://wiki.mozilla.org/FirefoxOS/New_security_model#Verifying_signatures_-_bug_1153422 and is otherwise not something web content can currently avail itself of.

The big wrinkle is that what is really needed is a cross-browser effort with this specific use-case in mind because it really is its own big picture idea.  I believe there are many people who care about the use-case, but I suspect most browser engines are focusing their efforts just on getting service workers and progressive web apps going.  The best hope for this use-case right now are browser extensions/add-ons.  All browsers seem to be converging on WebExtensions and at least some browsers (ex: Firefox) allow easy disabling of extension auto-updates.  Which is important since the trust-model looks like it depends on the extensions marketplaces' authentication mechanisms and them not being compromised themselves, which is strictly weaker than an air-gapped private key.  (NB: Pre-WebExtensions Firefox extensions can, however, be cryptographically signed.)  This is clearly nowhere as good as a packaged apps model that does not require installation like Firefox OS was shooting for, but it seems to be the most realistic cross-browser solution.

**Comment by twiss** [comment]
[Sorry for the 1.5 year late reply :) coming here from #1208, which is similar to this issue but instead of preventing the update, the old Service Worker warns the user about the update.]

@asutherland and others have brought up valid concerns about this approach, e.g., what if the user force-refreshes the web app, what if they open it in an incognito window, what about when they open it for the first time, what about on a new device. I agree with all of those, in fact, I [proposed a different solution](https://lists.w3.org/Archives/Public/public-webappsec/2017Apr/0042.html) built on Certificate Transparency a while back that would solve those things, and if browsers implemented that, I would be very happy too :)

However, a solution built on Service Workers has the big advantages that

1. It requires very little extra work on the part of browser developers
2. It potentially requires less work on the part of web app developers, as well. For example, instead of requiring them to sign their code with a public key, a SW solution can just require that updates be pushed to GitHub (or it can require both). A solution in the browser is unlikely to be flexible enough to allow that. And since many open source web apps already do that, it can be relatively more of a "set and forget" solution.

And I think it would solve the biggest part of this problem, especially because any attempt to send the user unsigned code runs the risk of detection by the SW, unless the server can somehow detect that the SW is being bypassed. That might be possible in the case of force-refresh (Cache-Control headers), though. Maybe we can use the SW to always send force-refresh-like Cache-Control headers?

And while some have indeed used browser extension to solve this problem (Signal, blockchain.info and mega.nz come to mind, although the first two of those are Chrome apps, which are getting phased out. I myself [made a Firefox addon](https://addons.mozilla.org/nl/firefox/addon/hcs-checker/)), it would be preferable if all users are protected, and they don't even protect against most of the loopholes ("what if they open it in an incognito window, what about when they open it for the first time, what about on a new device").

**Comment by MisterTicot** [comment]
Hi,

I want to secure my app by storing each asset SHA256SUM on a blockchain, then allow update when on-line file match the shasum. This way release have to be signed by author/auditors before update. This is an innovative use case of webworker/blockchain that can bring a new level of security to web applications, because an attacker would have to gain access to both the service and the private keys to perform a mass-attack.

However, this will be impossible to do as long as the service worker update is forced. Does someone knows about an alternate way to achieve this or eventually a feature implementation that could open this possibility?

I understand the issue about attacker being able to leverage the service worker self-control to implement malicious code durably. I'm not sure how this could be mitigated.

**Comment by MisterTicot** [comment]
It's been 3 month since I did propose a fix to this, and got no answer. Is there anything I can do so we can move on with this? Should I go ahead and make contact somewhere else? Or is it already settled ?

**Comment by asutherland** [comment]
This issue is on the upcoming ServiceWorkers F2F agenda for discussion in October, although I can't promise any resolution.

Note that my previous comments still stand about the security model of the web and in particular, the [Clear-Site-Data](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Clear-Site-Data) header since implemented by Chrome and Firefox has a "[kill switch](https://w3c.github.io/webappsec-clear-site-data/#example-killswitch)" which will wipe out all storage for an origin including ServiceWorkers.  That's an intentional feature that a SW would never be able to inhibit even if normal updates could be blocked or delayed.

There is a new experimental option that seems better suited to your use-case.  There's a **very** experimental Mozilla project https://github.com/mozilla/libdweb that enables WebExtensions to implement custom protocols like IPFS (see the [blog post](https://hacks.mozilla.org/2018/08/dweb-building-cooperation-and-trust-into-the-web-with-ipfs/)).


In regards to your proposal, I understand the core of your suggestion to be that a banner would be presented if the potentially-malicious SW blocks an update and that the potentially-malicious SW is able to provide some of the text to be displayed in the banner.  The user would then reach a decision informed by searching the web and asking other people on social media what they should do.

The issue with prompting the user in cases like this is that the user is frequently unable to make an informed decision about what is going on.  Especially if the attacker can supply scary text like "There is a virus in the update, don't update or your computer will be permanently damaged!" or text that leverages update fatigue like "The update will take 10 minutes to install, are you sure you want to update?" or unique strings that the attacker can use to game any search the user would make so that the guidance they find on the internet is from the attacker.  This isn't really solving the problem, it's making the problem the user's problem.

**Comment by MisterTicot** [comment]
Thank you for the answer and feeding the search.

My proposal has indeed nothing to do with bypassing clear browser cache functionality.

One of my premise was that in legit case, the banner wouldn't show for long as the server take-over is likely to be fixed within hours or at worth within days.

On the other hand, a malicious SW would continue to pop a warning forever until updated, incentivising user to do something about it. Non technical users are likely to accept the update after a few days and wipe malicious SW just to get rid of the banner. Not saying it's so great, just saying it leverages blind behaviors.

Now there may be a other ways to leverage this difference in update denial timespan, like setting a reasonable time limit to rejection at 24 or 48 hours.

The problem I see in their cases is that domain owners may know users have a malicious SW but remains powerless about it.

---

Another option (if it's acceptable for browsers)  I can see is having the SW rejection option being enabled for a website through a DNS TXT field.

In this scenario, the malicious SW squat attack against a website that hadn't the SW update rejection enabled could work only by taking over the DNS, and only as long as control over DNS remains as legit owner would be able to switch off the rejection option, triggering SW renewal.

The TXT field would be either inexistent (option not enabled) either a number that represent for how much minutes an update rejection remains valid, maybe with a reasonable maximum limit.

**Comment by jespertheend** [comment]
> I can't quite get my dead around the use cases.

The exact use case @mappum was describing seems to be what [is being discussed over here](https://community.signalusers.org/t/web-app-for-signal/1272/21). Although I can imagine not only E2EE apps require this. Almost any downloaded application uses codesigning to verify updates nowadays.
Servide Workers require https by default which I guess is the reason why code signed updates aren't really a thing on the web, and perhaps they're not really necessary. But I belive some applications could still benefit from this.

Perhaps being able to prevent server-forced updates is too generic. Maybe what we need here is something more specific to code signing updates. So that only Service Workers that make use of code signing may prevent server forced updates.
I'm not really sure how this would work though, because that way a malicious script could still block a Service Worker from getting updated by introducing code signing for the first time.

**Comment by hlandau** [comment]
Indeed, I previously proposed a solution for this problem, namely simply adding SRI support to service workers: https://github.com/w3c/webappsec-subresource-integrity/issues/66

Created an issue in the correct repository for discussion: #1680

**Comment by valpackett** [comment]
SRI cannot help this use case as SRI just uses hashes, not asymmetric signatures. SRI is only useful for cross-origin security.

---

I would like to implement signature verification in a service worker in my current project, where unlike in a typical web app, the client code is *user-controlled*; the user would upload a signed app update to the server, and the service worker would check the signature on the files coming from the server – allowing a trust-on-first-use model with regards to the server… if there was a "paranoia mode" for service workers that would disable all the escape hatches, making sure that the service worker could strip `Clear-Site-Data` and that update requests for the service worker would always go through it and never bypass it (and shift+reload would, I guess, ideally display a prompt about breaking the security? or just, like, get turned into a special message for the worker?).

Please, please, please give us **the choice**.

For the vast majority of web apps the current way of working, which would indeed "prevent an attacker from intentionally bricking the app forever", is the appropriate one.

For our paranoid E2EE apps with unconventional update models, we would like to **opt in** to full control of the update process with no escape hatches.

**Comment by hlandau** [comment]
>SRI cannot help this use case as SRI just uses hashes, not asymmetric signatures. SRI is only useful for cross-origin security.

Not true, as SRI can be used to specify an immutable bootloader which implements the asymmetric signature verification.

**Comment by valpackett** [comment]
Ah, that's clever. But either way, whether triggered by SRI or by passing some ` mode: 'paranoid' ` flag to `register()`, what's needed is a way to opt out of escape hatches.

**Comment by valpackett** [comment]
Oh, one key thing I've just realized: to not allow attackers to "permanently screw up" a compromised site that was using the normal mode (or no SWs at all), opting in to the secure mode should require at least a permission prompt, and probably be reflected in the browser UI (such as a "lock with refresh arrows" icon in place of the normal lock icon).

**Comment by JamesTheAwesomeDude** [comment]
While this doesn't solve the problem of *preventing* Service Worker updates, here's a rough sketch of an idea that might (with user co-operation) at least allow cobbling together a ToFU level of security to *detect* unauthorized updates, vet authorized updates, and fail-safe if an unvetted update occurs, on existing browsers:

- Call the actual "sensitive application data" (e.g. PGP private key) CS0.
- Each and every service worker is installed with a **unique per-installation** random secret, distributed as a "hard-coded constant" in the Service Worker's source code, wiped immediately from memory after generation (since at this time we're just aiming at ToFU security, *say* the server hasn't been compromised yet); call this CS1. This will fill a similar role to the [AEM machine validation secret](https://blog.invisiblethings.org/2011/09/07/anti-evil-maid.html#:~:text=if%20a%20user%20can%20see%20correct%20secret%20message%20(or%20perhaps%20a%20photo)%20being%20displayed%20on%20the%20screen%2C%20then%20it%20means%20that%20correct%20software%20must%20have%20booted), serving as living proof that the service worker is the same as it was when it was installed.
  * The endpoint that distributes the service workers must GENERALLY serve an HTTP 4XX response, except during active installation (which is indicated by an HTTPS-enabled cookie which is set by JS *immediately* before `register()` is called, and which is always unset by any non-4XX response from that endpoint)
- During "app setup" (Service Worker installation time), randomly generate a brief "challenge" passphrase and either generate or make the user generate a "client encryption" passphrase; call these P1 and PE.
- (Obviously, the service worker also does all the other things needed to keep a web application securely/immutably pinned.)
- Add Enc(CS1, CP) to LocalStorage. You need CS1 to reconstitute CP from LocalStorage.
- Add Enc(PE, Enc(CS1, CS0)) to LocalStorage.  You need PE *and* CS1 to reconstitute CS0 from LocalStorage.
  * For the love of god, please use at least half the available RAM for the KDF of the outer encryption layer here.
- Tell the user that they **MUST NOT EVER** disclose EP without verifying P1 first. Remind them that any request for PE, which is not accompanied by P1, is *certainly* from an adversary.
- Each time the application "starts up" (the page is visited), display P1 (maybe masked behind a clickwall with a reminder about shoulder-hoppers) and get EP input from the user; use that to decrypt CS0 into memory.
   * While the KDF is running, display a reminder to always validate P1 before entering EP
- When an application update is needed, the current version of the application follows this process:
   1. Fetch (using some API other than the service worker update URL) an alleged upcoming update of the service worker\*, which embeds pins for the alleged new version of the application
   2. Fetch and vet a hash of the alleged new version of the service worker\*. "Vetting" could include anything from checking it against a digital signature from the developer's offline signing keys, or multi-person geographically distributed [wrench](https://xkcd.com/538/)-resistant signatures, to checking multiple canaries, to getting the user to actually affirm its value.
      * \*Technically, this will just be a hash of "the service worker but with a dummy value for CS1".
   3. Once the new update's hash has been approved, execute this critical path subroutine:
      1. Assert that the *current* service worker has been installed with [`updateViaCache`](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/updateViaCache) to "all"
      2. Remind the user again that if they are prompted for EP again soon, that they **MUST NOT** disclose it unless P1 validates again
      3. Fetch a copy (from a **different** URL than the Service Worker's installation path) of the new service worker's source code
      4. Compare the hash of the new service worker's source code to the hash the user just approved
      5. If the hash compares unequal to the hash which the user had vetted, ABORT
      6. Modify the source code, replacing the embedded CS1 _\[some dummy value\]_ with the current value of CS1
      7. [Do an offline-update of the SW](https://stackoverflow.com/q/77233289/1874170), inserting the "modified" service worker that comprises validated source code plus device-specific value of CS1
      8. Restart the application (e.g. reload the page)

This should maybe be safe, since while an adversary can *overwrite* the existing service worker by simply serving a 200 OK response that would blow away CS (_cf._ an Evil Maid failing by blowing away TPM-protected secrets when factory-resetting the PC), I don't think there's any way for them to actually get at the source code of the *currently running* service worker.

The update mechanism is hella janky, though, and I'm less sure of it (especially since the "offline update" of a service worker isn't proven yet). The initial installation sort of *has* to include a leap of faith (that's ToFU), but I had hoped to allow the application to reduce the *surface area* the user has to think about to nothing more than matching an on-screen hash to an otherwise-known-good value. (Heck, you could even replace "user vetting" with )

-----

Obviously, this is a LOT of engineering work, and an additional UX burden of some song-and-dance. But, of course, at least *some* UX burden is absolutely necessary due to [this crux](https://github.com/w3c/ServiceWorker/issues/1680#issuecomment-1747606040). Certainly something more "batteries-included" for web applications to keep sensitive data safe from a surely upcoming adversarial takeover at an unknown future date would be nice.

**Comment by twiss** [comment]
Hi all :wave: FYI, I [proposed](https://github.com/WICG/proposals/issues/124) an alternative solution to the underlying goal (of facilitating web apps that don't trust the server) at the WICG (working title: [Source Code Transparency](https://github.com/twiss/source-code-transparency)). I also presented on it at the WebAppSec WG meeting at TPAC ([minutes](https://github.com/w3c/webappsec/blob/main/meetings/2023/2023-09-15-TPAC-minutes.md#source-code-transparency-sketch)), and it seemed like there was interest from the browsers there.

Instead of trying to prevent updates, the proposal here is to make all updates transparent and publicly auditable by security researchers, to make it detectable if any malicious code gets deployed by a web app's server. While this doesn't prevent malicious code from being deployed, it strongly discourages servers from ever doing so (due to the risk of reputational damage). The security model here is similar to [Certificate Transparency](https://certificate.transparency.dev/), which has been very successful at detecting and preventing malicious certificates from being issued. And contrary to the proposals here, it wouldn't be TOFU, but protect users from the first time they open the web app (if the browser implements source code transparency, obviously).

Even though I also previously [commented in favor of the proposal in this issue](https://github.com/w3c/ServiceWorker/issues/822#issuecomment-341866377), my impression is that browsers are quite resistant to preventing updates entirely, and would actually be more open to a solution dedicated to the underlying problem, rather than something "hacked" on top of Service Workers, even if it's more work to implement.

For the full proposal, please see the [explainer](https://github.com/twiss/source-code-transparency/blob/main/explainer.md). If you have any comments or suggestions, please open an issue or discussion on the [repo](https://github.com/twiss/source-code-transparency). If you support the proposal, please leave a :+1: on the [WICG proposal](https://github.com/WICG/proposals/issues/124). Thanks!

**Comment by jayaddison** [comment]
I'm interested in this functionality too, to protect disruption of the ServiceWorker within [RecipeRadar](https://www.reciperadar.com/), a web-hosted Progressive Web Application.

I've reviewed and broadly like @twiss's proposal, and I have a competing proposal that is compatible with delivery over both HTTP and HTTPS (my understanding is that SCT requires TLS).

My competing proposal is that site operators should deploy a [W3C-SRI-format compatible](https://www.w3.org/TR/SRI/) value to DNS containing the expected hash of the content body served from the root path of the webserver.  I admit that this single-path limitation is somewhat constraining.

Although I've asked the `dnsext` mailing list whether an additional DNS record type would be suitable for this, I think any kind of standardisation is a long way off, and perhaps unlikely.  However, for the sake of demonstration, I can illustrate a snippet of the current contents of DNS TXT records for `reciperadar.com`, where I've deployed the app's current expected hash:

```
;; ANSWER SECTION:
reciperadar.com.	3600	IN	TXT	"B=sha512-p69NIfGgc1xBwGBO+91ELEttAujx0vc4iCHJi+GnFLp1fdQGv8YXc1OXxky38vTwqzrG80FYySIvubNOPSGt4A=="
```

(note that when I deploy updated code for RecipeRadar, this entry temporarily contains _two_ hashes -- one for the cached/stale app allowing web browsers using web caches to continue to use the stale app until it expires, and one for the current/fresh app.  the W3C SRI spec foresaw this requirement for subresources hosted by CDNs, and so it does support multiple values at a given hash strength level - see [example 7 here](https://www.w3.org/TR/SRI/#does-response-match-metadatalist))

No web client currently supports this in practice, as far as I'm aware - however the idea would be that if the integrity check for the root resource fails, then we should be careful about trusting or loading any of the referenced subresources (including the ServiceWorker script), even if they contain SRI hashes (in other words: the root resource hash appears faulty, so all bets are off on the subresource hashes).

**Comment by jayaddison** [comment]
> I've reviewed and broadly like @twiss's proposal, and I have a competing proposal that is compatible with delivery over both HTTP and HTTPS (my understanding is that SCT requires TLS).

Sometimes I write comments too hastily.  To clarify: I've _informally_ reviewed SCT -- I'm not a member of any standards bodies, only a keen technologist -- and my proposal is to some extent competing, but there's no mutual exclusion between them (that is to say: both could be deployed in parallel).

**Comment by valpackett** [comment]
> deploy [SRI] to DNS

This might be really useful for concerns about a CDN becoming malicious while the legitimate operator still controls the DNS, but doesn't do anything for the "don't trust the operator" use case.

> similar to Certificate Transparency

Actually sounds compelling!… with the caveat that not every app is public and wants to be transparent, I guess.

---

Now that I think about it, what if we could have ServiceWorker/page-controlled updates under a special "installed web app" concept? So instead of imposing the controlled update model onto regular `https://` origins, there would be an action to "install" an app that would move it to a special `app://` origin that is basically a Cache-only one.. (For "add to home screen" this could happen implicitly, otherwise with a permission prompt for that in particular)

**Comment by jayaddison** [comment]
> This might be really useful for concerns about a CDN becoming malicious while the legitimate operator still controls the DNS, but doesn't do anything for the "don't trust the operator" use case.

Thanks @valpackett - yep, that's exactly the kind of scenario that a DNS webintegrity checksum would be intended to guard against; and correct, the mechanism does not protect against an untrusted operator.

(if an application _is_ free-and/or-open-source and reproducibly-buildable, then continuous inspection and confirmation of the published integrity hashes may be possible, but that'd be an independent process.  less-transparent sites could continue to offer content integrity)

I don't feel knowledgeable enough about either ServiceWorkers or web origins to comment on the `app://` origin suggestion, but to (try to) show some awareness: my understanding is that HTTPS is preferred for ServiceWorkers, so my proposal's attempt to design an approach that is backwards-compatible to HTTP could be out-of-context / off-topic here.

**Comment by stat3ofconsc3ience** [comment]
Bump?

**Comment by yoshisatoyanagisawa** [comment]
Let me summarize the discussion until here, regarding the proposal to allow Service Workers to opt out of server-forced updates.

The core of GitHub Issue #822 is a request to provide applications with a mechanism to prevent mandatory update checks, which currently occur at least every 24 hours. Proponents of this change argue that for security-sensitive applications, such as those handling private keys or end-to-end encryption (E2EE), the ability for a server to unilaterally push new code represents a significant attack surface. By controlling the update lifecycle, an application could verify the integrity and authenticity of new versions—for example, by checking for a cryptographic signature from the author—before allowing an update to proceed.

However, browser vendors have expressed serious concerns that allowing a Service Worker to defeat its own upgrades could be "an attacker's dream". If a malicious script were to gain control, it could use this feature to permanently persist on a user's device by blocking any subsequent legitimate fixes.

In light of these risks, the discussion has evolved toward more comprehensive solutions like Subresource Integrity (SRI) and Source Code Transparency (SCT). These efforts should be now being consolidated into the WAICT (Web Application Integrity, Consistency and Transparency) framework.

Given the existing arguments, solving this through the WAICT framework appears to be the most rational path forward, as it addresses both integrity and transparency in a unified manner. However, since WAICT is still in its early "Work in Progress" and standardization alignment phase, it seems premature to conclude on a specific implementation for Issue #822 at this stage. It would be more prudent to wait for the WAICT framework to mature and provide a standardized foundation for these high-trust requirements.

What do you think?

**Comment by valpackett** [comment]
@yoshisatoyanagisawa did you just copy-paste an LLM generated summary?!

**Comment by mitar** [comment]
@yoshisatoyanagisawa Thank you for the summary. I disagree that WAICT would solve everything here. I think probably a better approach would be to expose to end-users more information about service workers running in a background (maybe with an icon in the url bar, showing memory and process usage) with an option to kill it, remove it, etc. Also browsers could also kill and disable service workers for sites you do not visit often, etc.

I do not buy the argument that preventing updates from the server would allow attackers to gain an upper hand. In most cases server would not even know about a malicious service worker. And I am not even sure that I would want the server to silently "fix" the malicious service worker. I mean, if they managed to deploy a malicious service worker, I would expect the server to inform the user that there was a breach and that they are sorry and that you should do the following steps to remedy the situation.

**Comment by yoshisatoyanagisawa** [comment]
Since we are discussing a technical specification, I believe it is more appropriate to focus on systemic solutions rather than relying on the ethics or individual efforts of service providers. From an engineering perspective, security and reliability should be guaranteed by robust mechanisms that function automatically to protect the user.

Instead of relying on users to follow manual instructions after a breach, if a server needs to explicitly remediate a compromise, we require a well-defined recovery protocol and a mechanism that prevents attackers from interfering with those fixes. Do you have proposals for such a recovery mechanism?

Furthermore, just as most people use the web without understanding TCP/IP, they should be able to rely on a secure infrastructure without being forced to make technical decisions. Expecting users to understand and manually manage every Service Worker is an unrealistic proposal that compromises the accessibility and safety of the web for the general public.

**Comment by MicahZoltu** [comment]
Supply chain attacks and server compromises of high usage tools are now happening on a weekly, sometimes daily, basis.  I would like to see this revisited.

I can appreciate the desire to make it so that once an attacker gets onto a machine they aren't impossible to remove.

At the same time, I can also appreciate that trusting the server every time you open the app is a wildly different trust model from trusting the server when you first load the app.

Since these two problems are in direct conflict with each other, and both are legitimate, I also agree that we need a solution that can mitigate them both at the same time in some way.

Proposal:
Allow a registered service worker to *either* follow the current 24-hour update cycle, **OR** choose to prompt a user for every update.  In the latter case, the service worker would still check for updates every 24 hours, but when an update is available the browser would prompt the user and ask them if they would like to update.  Each update would only be prompted for once.

The idea here is that for low security apps, which are likely updating very regularly, the authors would follow the current setup where updates are automatic.  For high security apps the authors could choose to enable update prompting.  Their users would be "nagged" when updates are available, and thus developers of such apps would likely update less frequently and with more intent/fanfare (e.g., associated blog posts, etc.).

In the "prompt for update" regime, the default should probably be to *not* update, though this is certainly up for debate.  For a high security app in today's threat landscape my instinct is that the safe default when a user mindlessly clicks without reading is to not update rather than to update.

A design like this also could allow the browsers themselves to provide security warnings (if desired) when there is a known CVE with a particular version of a service worker deployed on the user's machine.  When they get the update prompt, the browser has the opportunity to warn the user about known CVEs.  The user would still be able to ignore the update, but it may give a path to deal with things like a spreading worm.

----

I believe this would address both the concern about attack service workers being "sticky" and difficult to get rid of, while also allowing for immutable apps that can be used when trust of server is very low and risk of compromise is very high.

**Comment by twiss** [comment]
FWIW, there is some parallel work happening on [Web App Integrity, Consistency and Transparency (WAICT)](https://waict.dev/) ([draft specs](https://github.com/waict-wg)) to address the underlying topic. Since this is being driven by Mozilla (among other parties) and has an experimental implementation in Firefox Nightly, it seems more viable as a path forward to me.

**Comment by MicahZoltu** [comment]
> FWIW, there is some parallel work happening on [Web App Integrity, Consistency and Transparency (WAICT)](https://waict.dev/) ([draft specs](https://github.com/waict-wg)) to address the underlying topic. Since this is being driven by Mozilla (among other parties) and has an experimental implementation in Firefox Nightly, it seems more viable as a path forward to me.

I don't think this solves the same problem.  With that proposal you are still trusting a remote third party to check the authenticity of your files.  At best, this turns a page into a 2/2 multisig scheme where an attacker needs to compromise both the verification file supplier and the server distributing the app.  This *is* strictly better than having 1/1, but not by much since a deployment compromise likely enables an attacker to deploy both the new app *and* update the verification file.

What I believe is being asked for here is the ability for a web app to have a trust model that is similar to installed software, where you trust the distributor at the time you download some installer/package, and then you never have to trust that distributor again.  You can frontload your authenticity verification at install time, checking with your extended trust network, auditing the code yourself if you have the ability to do so, etc. and then you can rely on that one-time verification to be durable over time.

Currently, the web largely doesn't support this, especially with `file://` being locked down and treated like an insecure context in many scenarios.  Modern browsers do not *allow* a user to do a one-time trust operation when using an app, they require a continuous trust model.  The only way to switch that to one-time is run your own localhost fileserver and download and serve apps from it.  This of course is not a reasonable user experience.

**Comment by thediaryofme** [comment]
Hello, I would be greatly interested in this. My website uses large pinned public keys and URL's. it would be absolutely wonderful for TOFU to be possible without a localhost install. Thank you! 🙏

**Comment by thediaryofme** [comment]
> > FWIW, there is some parallel work happening on [Web App Integrity, Consistency and Transparency (WAICT)](https://waict.dev/) ([draft specs](https://github.com/waict-wg)) to address the underlying topic. Since this is being driven by Mozilla (among other parties) and has an experimental implementation in Firefox Nightly, it seems more viable as a path forward to me.
> 
> I don't think this solves the same problem. With that proposal you are still trusting a remote third party to check the authenticity of your files. At best, this turns a page into a 2/2 multisig scheme where an attacker needs to compromise both the verification file supplier and the server distributing the app. This _is_ strictly better than having 1/1, but not by much since a deployment compromise likely enables an attacker to deploy both the new app _and_ update the verification file.
> 
> What I believe is being asked for here is the ability for a web app to have a trust model that is similar to installed software, where you trust the distributor at the time you download some installer/package, and then you never have to trust that distributor again. You can frontload your authenticity verification at install time, checking with your extended trust network, auditing the code yourself if you have the ability to do so, etc. and then you can rely on that one-time verification to be durable over time.
> 
> Currently, the web largely doesn't support this, especially with `file://` being locked down and treated like an insecure context in many scenarios. Modern browsers do not _allow_ a user to do a one-time trust operation when using an app, they require a continuous trust model. The only way to switch that to one-time is run your own localhost fileserver and download and serve apps from it. This of course is not a reasonable user experience.

Yes, this seems to work quite well. If the service worker could the user each update, it would prevent a hacked server from locking itself on the user, as well as prevent hackers from compromising a server. It seems to be less fragile.

**Comment by thediaryofme** [comment]
I think if this is to be done, browser controlled version freezing is best choice. Meaning, a "lock this page?" browser prompt that website code doesn't specifically trigger apart from by having a SW. This avoids malicious service workers from locking automatically. Thoughts? I actually dont think this belongs in the SW API at all: else it would be abused.

**Comment by MicahZoltu** [comment]
> I think if this is to be done, browser controlled version freezing is best choice. Meaning, a "lock this page?" browser prompt that website code doesn't specifically trigger apart from by having a SW. This avoids malicious service workers from locking automatically. Thoughts? I actually dont think this belongs in the SW API at all: else it would be abused.

That is an interesting idea.  I worry a bit about how to expose this to users in a way that is useful to them.  99% of users will just use defaults, so if it is off-by-default then most users are exposed to supply chain attacks for all of their service workers, and if it is on-by-default then most users are exposed to discovered security vulnerabilities.  The app developer likely knows the risk profile better (e.g., "I'm not maintaining this" or "I did 32 security audits of the code and none of the deployment key management process").  Of course, not all developers will make wise decisions either.

The whole situation seems to be full of tradeoffs, and while I'm pretty confident the current selected set of tradeoffs (not allowing any protection against supply chain attacks) is not ideal, it is not obvious to me which tradeoff set is "best" (it probably changes over time).

If we look to the future, supply chain attacks are on the rise and AI is making it easier to write software with fewer security vulnerabilities (when people actually put in the effort to do so), so the right tradeoff today may not be the right trade off in the future.  This makes me lean towards letting individual app/service worker developers set the policy so it can adapt over time, and making it easy for users to excise malicious service workers when they are notified of a problem.
